### PR TITLE
Fix imported model microscopic cross sections

### PR DESCRIPTION
**Pull request by amandalund**
This makes the lower limit on the model microscopic cross section grid consistent with Geant4 and only builds/imports the micro xs for models that use it.

WIP because I'm still confused about a few things:
- The Seltzer-Berger electron micro xs in stainless steel in four-steel-slabs is zero (this PR did not change those values, but I noticed when I was extending the `RootImporter` test). This seems wrong?
- Bremsstrahlung lambda tables are constructed using the secondary gamma production cut in the cross section calculation and as the lower energy grid bound, but for the micro xs the primary (e- or e+) production cut is used. Is this right?
- Not sure why my regenerated four-steel-slabs.root is breaking the `UrbanMsc` test.

### CODE REVIEW

**Comment by amandalund** [comment]
> Bremsstrahlung lambda tables are constructed using the secondary gamma production cut in the cross section calculation and as the lower energy grid bound, but for the micro xs the primary (e- or e+) production cut is used. Is this right?

I'm pretty sure the micro xs should be calculated using the secondary production cut, so I updated the importer accordingly. The brems micro xs still doesn't look right to me..

**Comment by amandalund** [comment]
Ok @sethrj @stognini I think this is ready now. In addition to using the secondary (instead of primary) production cut for the micro xs calculation and lower energy grid bound, I had to call `G4VEmModel::SetupForMaterial()`, which sets some energy- and material-dependent variables in the Geant4 model, before calling `ComputeCrossSectionPerAtom()`. The Seltzer-Berger and relativistic brems micro xs are now positive and energy-dependent (previously they were constant and in some cases zero),  and the upper SB xs values are approximately equal to the lower RB ones. I'd still feel a little better if we had some way to verify that we're not missing anything else..

**Comment by sethrj** [comment]
Yay! Since I now have geant4 pull request privileges I could try to push for giving *direct* access to the cross section tables in geant4.

File: src/celeritas/ext/detail/ImportProcessConverter.cc [hunk 1]
```diff
@@ -323,7 +332,13 @@ void ImportProcessConverter::store_em_tables(const G4VEmProcess& process)
         process_.models.push_back(model_class_id);
 
         // Save element-selector data
-        process_.micro_xs.insert({model_class_id, this->add_micro_xs(model)});
+        if (model_class_id == ImportModelClass::bethe_heitler_lpm
+            || model_class_id == ImportModelClass::klein_nishina
+            || model_class_id == ImportModelClass::livermore_rayleigh)
```
Comment (sethrj) [inline src/celeritas/ext/detail/ImportProcessConverter.cc:337]: Can you add a comment about why only these three models need the EM tables? Are you sure none of  the other muon (or other potential classes) need exporting?

File: src/celeritas/ext/detail/ImportProcessConverter.cc [hunk 2]
```diff
@@ -352,7 +367,13 @@ void ImportProcessConverter::store_energy_loss_tables(
         process_.models.push_back(model_class_id);
 
         // Save element-selector data
-        process_.micro_xs.insert({model_class_id, this->add_micro_xs(model)});
+        if (model_class_id == ImportModelClass::e_brems_sb
+            || model_class_id == ImportModelClass::e_brems_lpm
+            || model_class_id == ImportModelClass::mu_brems)
```
Comment (sethrj) [inline src/celeritas/ext/detail/ImportProcessConverter.cc:372]: Same here

File: src/celeritas/ext/detail/ImportProcessConverter.cc [hunk 3]
```diff
@@ -549,17 +570,22 @@ ImportProcessConverter::add_micro_xs(G4VEmModel& model)
...
+            = g4_cuts_table.GetMaterialCutsCouple(mat_id)->GetMaterial();
 
         ImportProcess::ElementPhysicsVectorMap element_physvec_map;
 
         // Set up all element physics vectors for this material and model
         for (const auto& elem_comp : material.elements)
         {
             ImportPhysicsVector physics_vector
-                = this->initialize_micro_xs_physics_vector(model, material);
+                = this->initialize_micro_xs_physics_vector(model, mat_id);
```
Comment (sethrj) [inline src/celeritas/ext/detail/ImportProcessConverter.cc:588]: Since this is an integer index rather than a `MatId`, maybe `mat_idx`?

File: src/celeritas/ext/detail/ImportProcessConverter.cc [hunk 4]
```diff
@@ -637,21 +674,38 @@ ImportProcessConverter::add_micro_xs(G4VEmModel& model)
...
-        {
-            min_energy = std::max(cutoff_iter->second.energy, min_energy);
-        }
+        const auto& g4_cuts_table
+            = *G4ProductionCutsTable::GetProductionCutsTable();
+        CELER_ASSERT(mat_id < g4_cuts_table.GetTableSize());
+        const auto g4_material
+            = g4_cuts_table.GetMaterialCutsCouple(mat_id)->GetMaterial();
+        const auto g4_particle
+            = G4ParticleTable::GetParticleTable()->FindParticle(
```
Comment (sethrj) [inline src/celeritas/ext/detail/ImportProcessConverter.cc:703]: These are pointer types I think, so better to make it clear that we're not making an unintentional copy:
```suggestion
        const auto* g4_material
            = g4_cuts_table.GetMaterialCutsCouple(mat_id)->GetMaterial();
        const auto* g4_particle
            = G4ParticleTable::GetParticleTable()->FindParticle(
```

File: src/celeritas/ext/detail/ImportProcessConverter.hh [hunk 5]
```diff
@@ -115,8 +115,7 @@ class ImportProcessConverter
 
     // Set up the physics vector energy grid for add_micro_xs(...)
     ImportPhysicsVector
-    initialize_micro_xs_physics_vector(G4VEmModel&           model,
-                                       const ImportMaterial& material);
+    initialize_micro_xs_physics_vector(G4VEmModel& model, unsigned int mat_id);
```
Comment (sethrj) [inline src/celeritas/ext/detail/ImportProcessConverter.hh:118]: But maybe it *would* be better to use a `MaterialId` type rather than an int...

File: src/celeritas/io/ImportProcess.hh [hunk 6]
```diff
@@ -128,6 +128,7 @@ struct ImportProcess
     //!@}
 
     int                                      particle_pdg;
+    int                                      secondary_pdg{0};
```
Comment (sethrj) [inline src/celeritas/io/ImportProcess.hh:131]: While you're at it, could you add default initializer to `particle_pdg` for safety?

File: test/celeritas/em/UrbanMsc.test.cc [hunk 7]
```diff
@@ -245,24 +245,19 @@ TEST_F(UrbanMscTest, msc_scattering)
...
     // Mock SimStateData
     SimStateValue states_ref;
     auto          sim_state_data = make_builder(&states_ref.state);
     sim_state_data.reserve(nsamples);
 
+    // Calculate range instead of hardcoding to ensure step and range values
+    // are bit-for-bit identical when range limits the step. The first three
+    // steps are not limited by range
+    std::vector<double> step = {0.00279169, 0.412343, 0.0376414};
+    step.reserve(nsamples);
```
Comment (sethrj) [inline test/celeritas/em/UrbanMsc.test.cc:259]: This and the logic below is a little hard to follow: how about
```suggestion
    constexpr double step_is_range = -1;
    std::vector<double> step = {0.00279169, 0.412343, 0.0376414};
    step.resize(nsamples, step_is_range);
```
and then below, instead of  `if (i == step.size()) push_back(...)`, `if (step[i] == step_is_range) step[i] = ...`?

Better, or worse?

File: test/celeritas/em/UrbanMsc.test.cc [hunk 8]
```diff
@@ -273,10 +268,18 @@ TEST_F(UrbanMscTest, msc_scattering)
...
+        {
+            PhysicsTrackView phys = this->make_track_view(
+                pdg::electron(), MaterialId{1}, MevEnergy{energy[i]});
+            step.push_back(phys.dedx_range());
+        }
     }
     const SimStateRef& states = make_ref(states_ref);
     SimTrackView       sim_track_view(states, ThreadId{0});
 
+    EXPECT_EQ(nsamples, step.size());
```
Comment (sethrj) [inline test/celeritas/em/UrbanMsc.test.cc:282]:
```suggestion
    ASSERT_EQ(nsamples, step.size());
```
to abort cleanly instead of segfaulting if the condition isn't met

**Review by sethrj — COMMENTED** [review]
Excellent! Thanks and sorry for the imported root data update :(

File: src/celeritas/ext/detail/ImportProcessConverter.cc (hunk 1 above)
Comment (amandalund) [reply src/celeritas/ext/detail/ImportProcessConverter.cc:337]: Good point, there are a couple planned models (eCoulombScattering, muPairProduction) that build element selectors -- i'll add those too.

File: test/celeritas/em/UrbanMsc.test.cc (hunk 7 above)
Comment (amandalund) [reply test/celeritas/em/UrbanMsc.test.cc:259]: better :)

File: src/celeritas/ext/detail/ImportProcessConverter.cc (hunk 1 above)
Comment (sethrj) [reply src/celeritas/ext/detail/ImportProcessConverter.cc:337]: Since we're not able to transport protons in celeritas, then if we want to add the coulomb scattering process, the production cutoff should be effectively infinite, right? (I.e., protons of any energy get deposited locally via dE/dx.) Is there a way to get Geant4 to do this?

Comment (amandalund) [reply src/celeritas/ext/detail/ImportProcessConverter.cc:337]: Deferring to #485

File: src/celeritas/ext/detail/ImportProcessConverter.cc [hunk 9]
```diff
@@ -137,7 +138,7 @@ ImportModelClass to_import_model(const G4VEmModel& model)
         {"eBremLPM",            ImportModelClass::e_brems_lpm},
         {"eplus2gg",            ImportModelClass::e_plus_to_gg},
         {"LivermorePhElectric", ImportModelClass::livermore_photoelectric},
-        {"KleinNishina",        ImportModelClass::klein_nishina},
+        {"Klein-Nishina",       ImportModelClass::klein_nishina},
```
Comment (sethrj) [inline src/celeritas/ext/detail/ImportProcessConverter.cc:141]: Geant4 is the worst...

**Review by sethrj — APPROVED** [review]
Yay! Thanks for crushing this. 💪

File: src/celeritas/ext/detail/ImportProcessConverter.cc (hunk 9 above)
Comment (amandalund) [reply src/celeritas/ext/detail/ImportProcessConverter.cc:141]: Clearly the "-" stands for "minus atomic shell effects"

**Comment by sethrj** [comment]
CI is having issues but I double checked that the debug (non-vecgeom) build passes.
